### DISCUSSION TITLE
extract asset from bundle and save it

### DISCUSSION THREAD

**Discussion by karlito31** [nesrak1/AssetsTools.NET#70] [Q&A] [answered]
Is it possible to extract the asset from the bundle and save as a separate asset file or in new (empty?) bundle, without changing anything in the asset itself? If so, how would I achieve that? Runtime, not in editor.
The second question, a little more complicated, is it possible to save an instantiated GameObject from a bundle as an asset file?
Thanx.

**Comment by nesrak1** [comment]
> and save as a separate asset file

BundleHelper.LoadAssetDataFromBundle can give you a byte array from a bundle

> or in new (empty?) bundle

If you want to import into a new empty bundle, there's no way to generate one right now easily (there are slight variations between engine versions and they're laid out differently). If you still want to do it, I've told people to build a bundle with the same engine version as the game in the editor and then delete all the files with AssetsTools.NET or UABEA. That would give you a blank bundle with no files. Then you can open the bundle and immediately write it in AssetsTools.NET (using a BundleReplacer) to make a bundle with only that file. You should also pass the class database file (assetsManager.classFile) into the bundle write method to generate a type tree.

**Comment by karlito31** [comment, marked as answer]
So, If I want to extract particular GameObject from bundle (with all its dependencies), I can do that with BundleHelper.LoadAssetDataFromBundle ? And than write that bytearray to a new asset file (or new prepared empty bundle, as per your answer) ?
Sorry, its all new to me :)

**Reply by nesrak1** [reply]
Yeah, LoadAssetDataFromBundle reads the entire assets file you choose into a byte array.

> So, If I want to extract particular GameObject from bundle (with all its dependencies)

It will extract all the assets in that file which probably include the gameobject you're looking for among other assets. As for dependencies, it completely depends.

If it's a data.unity3d bundle then it 99% has dependencies and you will have to also export the dependencies yourself (check assetsFileInstance.file.dependencies recursively). Otherwise, bundles are generally not dependent on other assets so you wouldn't have to worry about it.

**Comment by karlito31** [comment]
Ok tnx, will try tomorrow. I might be back for additional help.
